crates/zkvm-interface/platform/src/lib.rs

```diff
@@ -54,3 +59,73 @@ pub trait Platform {
         t
     }
 }
+
+/// Wrapper for `Platform` implementation that hashes output before calling
+/// the inner `P::write_whole_output`.
+pub struct OutputHashedPlatform<P, D>(PhantomData<(P, D)>);
+
+impl<P, D> Platform for OutputHashedPlatform<P, D>
+where
+    P: Platform,
+    D: Digest,
+{
+    #[inline]
+    fn read_whole_input() -> impl Deref<Target = [u8]> {
+        P::read_whole_input()
+    }
+
+    #[inline]
+    fn write_whole_output(output: &[u8]) {
+        P::write_whole_output(&D::digest(output));
+    }
+
+    #[inline]
+    fn print(message: &str) {
+        P::print(message);
+    }
+
+    #[inline]
+    fn cycle_count() -> u64 {
+        P::cycle_count()
+    }
+
+    #[inline]
+    fn cycle_scope_start(name: &str) {
+        P::cycle_scope_start(name)
+    }
+
+    #[inline]
+    fn cycle_scope_end(name: &str) {
+        P::cycle_scope_end(name)
+    }
+
+    #[inline]
+    fn cycle_scope<T>(name: &str, f: impl FnOnce() -> T) -> T {
+        P::cycle_scope(name, f)
+    }
+}
+
+/// Stdin with a LE u32 length prefix.
+///
+/// Dereferencing it returns slice to the actual data.
+pub struct LengthPrefixedStdin(Vec<u8>);
+
+impl Deref for LengthPrefixedStdin {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0[4..]
+    }
+}
+
+impl LengthPrefixedStdin {
+    pub fn new(stdin: Vec<u8>) -> Self {
+        let len = u32::from_le_bytes(stdin[..4].try_into().unwrap());
```

Comment:
Maybe paranoid comment but if `stdin` is shorter than 4 bytes, we'll panic. I think since this is internal it _shouldn't_ happen. Not sure if might be annoying to retun a `Result` -- no strong opinions.

Edit: saw `with_prefixed_stdin` below -- I guess if this is the only use should be okay.

Reply:
> Uhm, this can also panic if whateverzkvmsdk::read_input() is shorter than 4 bytes -- but on properly used API this shouldn't happen. Maybe just worth an assert with a nicier panic msg, just to help someone misusing the api.

Yeah this makes sense, added some more panic message in https://github.com/eth-act/ere/pull/235/commits/a9f048966b6592f54c1a20281e0c8888a86900df.